python/ray/autoscaler/gcp/node_provider.py

```diff
@@ -0,0 +1,194 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from googleapiclient import discovery
+
+from ray.autoscaler.node_provider import NodeProvider
+from ray.autoscaler.gcp.config import wait_for_compute_zone_operation
+from ray.autoscaler.gcp.tags import TAG_KEYS, TAG_VALUES
+from ray.ray_constants import BOTO_MAX_RETRIES
+
+TERMINATED_STATES = (
+    'STOPPING',
+    'STOPPED',
+    'SUSPENDING',
+    'SUSPENDED',
+    'TERMINATED',
+)
+
+class GCPNodeProvider(NodeProvider):
+    def __init__(self, provider_config, cluster_name):
+        NodeProvider.__init__(self, provider_config, cluster_name)
+
+        self.compute = discovery.build('compute', 'v1')
+
+        self.tag_keys = TAG_KEYS.copy()
+        self.tag_values = TAG_VALUES.copy()
+
+        # Cache of node objects from the last nodes() call. This avoids
+        # excessive DescribeInstances requests.
+        self.cached_nodes = {}
+
+        # Cache of ip lookups. We assume IPs never change once assigned.
+        self.internal_ip_cache = {}
+        self.external_ip_cache = {}
+
+    def nodes(self, tag_filters):
+        if tag_filters:
+            label_filter_expr = '(' + ' AND '.join([
+                '(labels.{key} = {value})'.format(key=key, value=value)
+                for key, value in tag_filters.items()
+            ]) + ')'
+        else:
+            label_filter_expr = ''
+
+        instance_state_filter_expr = '(' + ' OR '.join([
+            '(status = {status})'.format(status=status)
+            for status in ['PROVISIONING', 'STAGING', 'RUNNING']
+        ]) + ')'
+
+        cluster_name_filter_expr = (
+            '(labels.{key} = {value})'
+            ''.format(key=self.tag_keys['cluster-name'],
+                      value=self.cluster_name))
+
+        not_empty_filters = [
+            f for f in [
+                    label_filter_expr,
+                    instance_state_filter_expr,
+                    cluster_name_filter_expr,
+            ] if f
+        ]
+
+        filter_expr = ' AND '.join(not_empty_filters)
+
+        response = self.compute.instances().list(
+            project=self.provider_config['project_id'],
+            zone=self.provider_config['availability_zone'],
+            filter=filter_expr,
+        ).execute()
+
+        instances = response.get('items', [])
+        # Note: All the operations use 'name' as the unique instance identifier
```

Comment:
I assume this is guaranteed to be unique per node?

Reply:
Yep. In fact, I had to change the [create_node](https://github.com/ray-project/ray/pull/2061/files#diff-6a987ff43a9906e4d9d2116bc57c22c7R464) arguments to pass unique name to the nodes because gcp requires the names to be unique.